### PR TITLE
353 render UI conditionally based on use rights

### PR DESCRIPTION
**Pull request by PhilKsr**
DESCRIPTION
- add rights check utils function
- save rights to store
- check on any action for required rights

CLOSES
Related to https://github.com/Frachtwerk/essencium-frontend/issues/353

### CODE REVIEW

File: packages/lib/src/components/NavBar/components/NavLinks.tsx [hunk 1]
```diff
@@ -29,22 +29,18 @@ import { useTranslation } from 'next-i18next'
...
 
   const theme = useMantineTheme()
 
   function hasRequiredRights(rights: string[]): boolean {
-    return Boolean(
-      rights.every(right =>
-        user?.role.rights.map(userRight => userRight.authority).includes(right)
-      )
-    )
+    return Boolean(rights.every(right => userRights?.includes(right)))
```
Comment (qhantom) [inline packages/lib/src/components/NavBar/components/NavLinks.tsx:43]: `.every` already does return a boolean hence it needn't be explicitly coerced. But you already have written a util function `hasRequiredRights` so you can use it here right?

EDIT: Ah, I've seen that the util function is only in the `app` package. Suggestion: outsource it to the `lib` package and import it in the `app` package like other util functions.

File: packages/app/src/pages/translations/index.tsx [hunk 2]
```diff
@@ -48,6 +51,8 @@ function getTranslationsByLanguage(
 function TranslationsView(): JSX.Element {
   const { t } = useTranslation()
 
+  const [userRights] = useAtom(userRightsAtom)
```
Comment (qhantom) [inline packages/app/src/pages/translations/index.tsx:54]:
```suggestion
  const userRights = useAtomValue(userRightsAtom)
```

File: packages/app/src/pages/translations/index.tsx [hunk 3]
```diff
@@ -118,6 +123,8 @@ function TranslationsView(): JSX.Element {
       </Title>
 
       <Translations
+        canDelete={hasRequiredRights(userRights, RIGHTS.TRANSLATION_DELETE)}
+        canUpdate={hasRequiredRights(userRights, RIGHTS.TRANSLATION_UPDATE)}
```
Comment (qhantom) [inline packages/app/src/pages/translations/index.tsx:127]: Why can't we put the rights-logic into the `Translations.tsx` file (`lib`) and use props here?

File: packages/app/src/pages/users/index.tsx [hunk 4]
```diff
@@ -97,6 +98,8 @@ function UsersView(): JSX.Element {
 
   const [user] = useAtom(userAtom)
 
+  const [userRights] = useAtom(userRightsAtom)
```
Comment (qhantom) [inline packages/app/src/pages/users/index.tsx:101]:
```suggestion
  const user = useAtomValue(userAtom)

  const userRights = useAtomValue(userRightsAtom)
```

File: packages/app/src/pages/roles/index.tsx [hunk 5]
```diff
@@ -83,7 +84,7 @@ function RolesView(): JSX.Element {
 
   const theme = useMantineTheme()
 
-  const [user] = useAtom(userAtom)
+  const [userRights] = useAtom(userRightsAtom)
```
Comment (qhantom) [inline packages/app/src/pages/roles/index.tsx:87]:
```suggestion
  const userRights = useAtomValue(userRightsAtom)
```

File: packages/app/src/pages/rights/index.tsx [hunk 6]
```diff
@@ -51,18 +51,19 @@ import { useAtom } from 'jotai'
...
+import { hasRequiredRights } from '@/utils/hasRequiredRights'
 import { baseGetStaticProps } from '@/utils/next'
 
 const DEFAULT_SORTING: SortingState = [{ id: 'authority', desc: false }]
 
 function RightsView(): JSX.Element {
   const { t } = useTranslation()
 
-  const [user] = useAtom(userAtom)
+  const [userRights] = useAtom(userRightsAtom)
```
Comment (qhantom) [inline packages/app/src/pages/rights/index.tsx:66]:
```suggestion
  const userRights = useAtomValue(userRightsAtom)
```

**Review by qhantom — CHANGES_REQUESTED** [review]
Looks nice! What is still missing is the route protection. If a user does not have the appropriate rights, he shouldn't even be able to get to the page by navigating to it directly:

image

**Comment by qhantom** [comment]
@PhilKsr I saw, that I can update a rights without having the `RIGHT_UPDATE` right but only the `ROLE_UPDATE` right.

File: packages/lib/src/components/NavBar/components/NavLinks.tsx (hunk 1 above)
Comment (PhilKsr) [reply packages/lib/src/components/NavBar/components/NavLinks.tsx:43]: Very good suggestion!

**Review by qhantom — CHANGES_REQUESTED** [review]
Looks better! But I can still access the route without having the rights:

image

**Comment by PhilKsr** [comment]
> Looks better! But I can still access the route without having the rights:
> 
> image

Yes, because I wasn't done yet. xD

@qhantom FYI: This is a very basic approach. Otherwise it's not an S ticket anymore. I think we should create an own issue for routeguards, including a wildcard. What do you think?

**Comment by qhantom** [comment]
> > Looks better! But I can still access the route without having the rights:
> > image
> 
> Yes, because I wasn't done yet. xD
> 
> @qhantom FYI: This is a very basic approach. Otherwise it's not an S ticket anymore. I think we should create an own issue for routeguards, including a wildcard. What do you think?

I get an endless loop of requests when looking at the network tab. Since it was labeled `S` and takes longer, we can transfer it to me if you're fine with it and if you don't have no more capacity.

**Comment by PhilKsr** [comment]
> > > Looks better! But I can still access the route without having the rights:
> > > image
> > 
> > 
> > Yes, because I wasn't done yet. xD
> > @qhantom FYI: This is a very basic approach. Otherwise it's not an S ticket anymore. I think we should create an own issue for routeguards, including a wildcard. What do you think?
> 
> I get an endless loop of requests when looking at the network tab. Since it was labeled `S` and takes longer, we can transfer it to me if you're fine with it and if you don't have no more capacity.

Dammit! Yes it would be fine for me. You have the capacity?

**Comment by qhantom** [comment]
> > > > Looks better! But I can still access the route without having the rights:
> 
> > > > image
> 
> > > 
> 
> > > 
> 
> > > Yes, because I wasn't done yet. xD
> 
> > > @qhantom FYI: This is a very basic approach. Otherwise it's not an S ticket anymore. I think we should create an own issue for routeguards, including a wildcard. What do you think?
> 
> > 
> 
> > I get an endless loop of requests when looking at the network tab. Since it was labeled `S` and takes longer, we can transfer it to me if you're fine with it and if you don't have no more capacity.
> 
> 
> 
> Dammit! Yes it would be fine for me. You have the capacity?

Yep, gonna have a look at it.

**Comment by PhilKsr** [comment]
@qhantom please have a final look. Do you know why the Vercel deployment fails after rebase? Probably issues with the force push.

**Review by qhantom — APPROVED** [review]
Nice! 🚀

**Pull request merged**